aea/protocols/dialogue/base.py

```diff
@@ -857,66 +851,119 @@ def _basic_validation_non_initial_message(
+            latest_ids.append(abs(self.last_incoming_message.message_id))
+
+        if self.last_outgoing_message:
+            latest_ids.append(abs(self.last_outgoing_message.message_id))
+
+        if abs(target) > max(latest_ids):
+            return "Invalid target. Expected a value less than or equal to abs({}). Found abs({}).".format(
+                self._last_message_id, target
             )
 
-        target_message = self._get_message(target)
+        # detailed target check
+        target_message = self.get_message_by_id(target)
+
+        if not target_message:
+            return "Invalid target {}. target_message can not be found.".format(target)
+
         target_performative = target_message.performative
         if performative not in self.rules.get_valid_replies(target_performative):
-            return (
-                False,
-                "Invalid performative. Expected one of {}. Found {}.".format(
-                    self.rules.get_valid_replies(target_performative), performative
-                ),
+            return "Invalid performative. Expected one of {}. Found {}.".format(
+                self.rules.get_valid_replies(target_performative), performative
             )
 
-        return True, "The non-initial message passes basic validation."
+        return None
 
-    def _additional_validation(self, message: Message) -> Tuple[bool, str]:
-        """
-        Check whether 'message' is a valid next message in the dialogue, according to additional rules.
+    def _validate_message_id(self, message: Message) -> Optional[str]:
+        """Check message id corresponds to message id sequences, if not return error string."""
+        is_outgoing = message.to != self.self_address
 
-        These rules are designed to be less fundamental than basic rules and subject to change.
-        Currently the following is enforced:
+        # This assumes that messages sent by the opponent are sent in the right order.
+        if is_outgoing:
+            next_message_id = self.get_outgoing_next_message_id()
+        else:
+            next_message_id = self.get_incoming_next_message_id()
 
-         - A message targets the message strictly before it in the dialogue
+        # we know what is the next message id for incoming and outgoing!
+        if message.message_id != next_message_id:
+            return "Invalid message_id. Expected {}. Found {}.".format(
+                next_message_id, message.message_id
+            )
 
-        :param message: the message to be validated
-        :return: Boolean result, and associated message.
-        """
-        if self.is_empty:
-            return True, "The message passes additional validation."
+        return None
 
-        last_target = self.last_message.target  # type: ignore
-        if message.target == last_target + 1:
-            return True, "The message passes additional validation."
+    def iter_all_messages(self) -> Iterable:
```

Comment:
There are places where `chain(self._incoming_messages, self._outgoing_messages)` is used, yet this method encapsulates this. Why not use this method instead?

Reply:
Although the place it is used it is not needed as per my comment. If removed there and this method isn't used, remove this method.

Reply:
removed, thanks